vyper/semantics/analysis/annotation.py

```diff
@@ -157,10 +175,30 @@ def visit_Call(self, node, type_):
             ):
                 self.visit(value, arg_type)
         elif isinstance(call_type, MemberFunctionT):
+            if call_type.is_modifying:
+                # it's a dotted function call like dynarray.pop()
+                expr_info = get_expr_info(node.func.value)
+                expr_info.validate_modification(node, self.func.mutability)
+
             assert len(node.args) == len(call_type.arg_types)
             for arg, arg_type in zip(node.args, call_type.arg_types):
                 self.visit(arg, arg_type)
         else:
+            # note that mutability for`raw_call` is handled in its `build_IR` function
+            mutable_builtins = (
+                "create_minimal_proxy_to",
+                "create_copy_of",
+                "create_from_blueprint",
+            )
```

Comment:
There is an overlap here with https://github.com/vyperlang/vyper/pull/3546#discussion_r1284954648.

Reply:
Another option here could be to check constancy of the context in `_CreateBase.build_IR()`, similar to how it is being checked for `raw_call`. However, this would not resolve #3546.